lib/StringClient.js

```diff
@@ -182,72 +184,114 @@ StringClient.prototype.write = function write(options, body, callback) {
+        hash = crypto.createHash('md5');
+    }
 
-        if (res) {
-            md5 = res.headers['content-md5'];
+    if (res.headers['content-encoding'] === 'gzip') {
+        decoder = new StringDecoder('utf8');
+        gz = zlib.createGunzip();
+        gz.on('data', function (chunk) {
+            body += decoder.write(Buffer.from(chunk));
+        });
+        gz.once('end', function () {
+            body += decoder.end();
+            done();
+        });
+        res.once('end', gz.end.bind(gz));
+    } else {
+        res.setEncoding('utf8');
+        res.once('end', done);
+    }
 
-            if (md5 && req.method !== 'HEAD' && res.statusCode !== 206) {
-                hash = crypto.createHash('md5');
-            }
+    res.on('data', function onData(chunk) {
+        if (hash) {
+            hash.update(chunk);
+        }
 
-            if (res.headers['content-encoding'] === 'gzip') {
-                decoder = new StringDecoder('utf8');
-                gz = zlib.createGunzip();
-                gz.on('data', function (chunk) {
-                    body += decoder.write(Buffer.from(chunk));
-                });
-                gz.once('end', function () {
-                    body += decoder.end();
-                    done();
-                });
-                res.once('end', gz.end.bind(gz));
-            } else {
-                res.setEncoding('utf8');
-                res.once('end', done);
-            }
+        if (gz) {
+            gz.write(chunk);
+        } else {
+            body += chunk;
+        }
+    });
+};
 
-            res.on('data', function onData(chunk) {
-                if (hash) {
-                    hash.update(chunk);
-                }
 
-                if (gz) {
-                    gz.write(chunk);
-                } else {
-                    body += chunk;
-                }
-            });
+/**
+ * upon a successful result from underlying HttpClient, this function is called.
+ * parse the response, return the callback, and emit an after event. return a
+ * function to bind the callback to the closure scope.
+ * @private
+ * @method _onResult
+ * @param {Function} callback user land callback provided to client verb method
+ * @returns {Function}
+ */
+StringClient.prototype._onResult = function _onResult(callback) {
 
-        } else {
-            callback(resErr, req, null, null);
+    var self = this;
+
+    // this function is invoked by HttpClient's 'result' event
+    return function onResult(err, res, req) {
+        // A bit tricky, we can't just do a check on err object here:
+        // Scenario 1: we have an err here, but no res object, e.g., req timeout
+        // Scenario 2: we got an http 500 err, along with res object
+        // Instead, let's check for res object instead. If it doesn't exist,
+        // it's scenario 1 and we can return right away.
+        if (!res) {
+            callback(err, req, null, null);
+            self.emit('after', req, null, err);
```

Comment:
If we're only emitting 'after' in the string client, doesn't this mean the raw HTTP client won't emit an after event?